ex2_hal/adcs/equipment_handler/src/adcs_io.c

```diff
@@ -151,36 +162,54 @@ ADCS_returnState request_uart_telemetry(uint8_t TM_ID, uint8_t *telemetry, uint3
         return ADCS_UART_BUSY;
     }
 
-    uint8_t frame[5];
+    // Form the command frame
+    uint8_t frame[ADCS_TM_HEADER_SZ];
     frame[0] = ADCS_ESC_CHAR;
     frame[1] = ADCS_SOM;
     frame[2] = TM_ID;
     frame[3] = ADCS_ESC_CHAR;
     frame[4] = ADCS_EOM;
 
-    sciSend(ADCS_SCI, 5, frame);
+    // Send the command frame
+    sciSend(ADCS_SCI, ADCS_TM_HEADER_SZ, frame);
     if (xSemaphoreTake(tx_semphr, UART_TIMEOUT_MS) != pdTRUE) {
         return ADCS_UART_FAILED;
     }
 
     int received = 0;
-    uint8_t *reply = (uint8_t *)pvPortMalloc(length + 5);
-    if (reply == NULL) {
-        return ADCS_MALLOC_FAILED;
-    }
+    uint8_t *reply = (uint8_t *)pvPortMalloc(length + ADCS_TM_HEADER_SZ);
+    if (reply == NULL) return ADCS_MALLOC_FAILED;
 
-    while (received < length + 5) {
+    bool end_of_message = false;
+    const uint8_t ending_bytes[ADCS_NUM_ENDING_BYTES] = {ADCS_PARSING_BYTE, ADCS_ENDING_BYTE};
+    uint8_t ending_bytes_index = 0;
+
+    while (!end_of_message) {
         if (xQueueReceive(adcsQueue, reply + received, UART_TIMEOUT_MS) == pdFAIL) {
             return ADCS_UART_FAILED;
```

Comment:
missing free of *reply

Reply:
Happens near the end of this function, since it's still being used up until that point

Reply:
Doesn't happen at the end of the function if it returns here!

Reply:
fax